paper_framework_sources/p2079_system_execution_context.bs

```diff
@@ -504,25 +574,18 @@ Underneath the `system_context` there is a singleton of some sort, potentially o
 The question is how we expose the singleton.
 We have a few obvious options:
 
- - Explicit context objects, as we've described in R2 and R3 of this paper, where a `system_context` is constructed as any other context might be, and refers to a singleton underneath.
+ - Explicit context objects, as we've described in R2, R3 and R4 of this paper, where a `system_context` is constructed as any other context might be, and refers to a singleton underneath.
  - A global `get_system_context()` function that obtains a `system_context` object, or a reference to one, representing the singleton explicitly.
  - A global `get_system_scheduler()` function that obtains a scheduler from some singleton system context, but does not explicitly expose the context.
 
-The `get_system_context()` function returning by value adds little, it's equivalent to direct construction.
-`get_system_context()` returning by reference and `get_system_scheduler()` have a different lifetime semantic from directly constructed `system_context`.
-
-The main reason for having an explicit by-value context is that we can reason about lifetimes.
-If we only have schedulers, from `get_system_context().get_scheduler()` or from `get_system_scheduler()` then we have to think about how they affect the context lifetime.
-We might want to reference count the context, to ensure it outlives the schedulers, but this adds cost to each scheduler use, and to any downstream sender produced from the scheduler as well that is logically dependent on the scheduler.
-We could alternatively not reference count and assume the context outlives everything in the system, but that leads quickly to shutdown order questions and potential surprises.
-
-By making the context explicit we require users to drain their work before they drain the context.
-In debug builds, at least, we can also add reference counting so that destruction of the context before work completes reports a clear error to ensure that people clean up.
-That is harder to do if the context is destroyed at some point after main completes.
-This lifetime question also applies to construction: we can lazily construct a thread pool before we first use a scheduler to it.
-
-For this reason, and consistency with other discussions about structured concurrency, we opt for an explicit context object here.
+In R4 and previous revisions, we opted for the explicit context object.
+The argument was that providing explicit contexts is easier to reason about the lifetime of the schedulers.
+However, reasoning for the lifetime of schedulers (and the operations spawned on them) may not be that useful.
+Ultimatelly, all the operations launched on schedulers need to be contained in the lifetime of `main()`.
+Adding an artificial scope for reasoning about lifetimes is just an extra burden.
```

Comment:
People are likely to have concerns about very visible singleton-like nature of this thing. We can call get_system_scheduler() from anywhere and it's going to be extremely easy to slip up on lifetimes. Is it reasonable to expect runtimes to do lifetime checking? Can we even in production builds ensure that tasks error out after main?